### ISSUE TITLE
Issued while checking for update of Mulch

### ISSUE THREAD

**Issue by Universalizer** [Tobi823/ffupdater#262]
### **Thanks for your app, but Issued while checking for update of Mulch, as follows :-**

Stacktrace:
```
java.lang.Exception: The app does not support the device
	at de.marmaro.krt.ffupdater.device.DeviceAbiExtractor.findBestAbiForDeviceAndApp(Unknown Source:41)
	at de.marmaro.krt.ffupdater.app.impl.Mulch.findLatestUpdate$ffupdater_release(SourceFile:65)
	at de.marmaro.krt.ffupdater.app.impl.AppBase.findAppUpdateStatus(Unknown Source:64)
	at de.marmaro.krt.ffupdater.app.impl.AppBase.access$findAppUpdateStatus(Unknown Source:0)
	at de.marmaro.krt.ffupdater.app.impl.AppBase$checkForUpdateAsync$2$1.invokeSuspend(Unknown Source:115)
	at w3.a.resumeWith(Unknown Source:8)
	at j4.i0.run(Unknown Source:100)
	at m4.d.run(Unknown Source:11)
	at n4.j.run(Unknown Source:2)
	at n4.a$a.run(SourceFile:82)
```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.3.1 (115) release |
| Device | SM-J600G (j6ltedx, j6lte, exynos7870) |
| Manufacturer | samsung (samsung) | 
| Supported ABIs | armeabi-v7a, armeabi |
| Android version | 10 (SDK: 29) |
| OS | Android 32-bit Only (Not 64-bit)  |         

_By the way can you, “allow user’s to send feedback within your app’s menu”, so that direct reporting can be possible._

**Comment by Tobi823** [comment]
First issue: ups, I missed that Mulch supports arm64 and armeabi-v7a

**Comment by Tobi823** [comment]
Second issue: Thanks for the update. I was not aware that Mulch is also available on F-Droid.

**Comment by Tobi823** [comment]
> still it does not appear in the list of FF Updater,

Because FFUpdater will only list installed apps with the correct signed key / certificate fingerprint. If it does not match, then the user can't update the app and he/she will be confused

**Comment by Tobi823** [comment]
>  it downloads "64bit armeabi-v8a version code file (false and non supportive)", even on successful download but fail to install or update, then display some errors notification.

will be fixed in the latest version

**Comment by Tobi823** [comment]
> But (it doesn't) Appears and as if “Add New Installation” is required (Noted ; which is already Installed),

I will add a warning if the fingerprint of the installed app does not match. All non matching apps will be ignored by FFUpdater.

![Screenshot_20221102_231621](https://user-images.githubusercontent.com/10066140/199613517-79663286-83c8-4f93-807d-62fb54481e10.jpg)

**Comment by Tobi823** [comment]
done, 77.3.4 is now released

**Comment by Tobi823** [comment]
Sorry for me late response. The bug should be really fixed in 77.4.0

**Comment by Tobi823** [comment]
Many devices support 32 bit and 64 bit apps like Windows. On Windows 10 64bit you can still run older 32 bit applications.

With SUPPORTED_ABIS you can query which ABIs are supported on your device (https://developer.android.com/reference/android/os/Build#SUPPORTED_ABIS). The ABI listed first in SUPPORTED_ABIS is usually better supported than the last ABI in SUPPORTED_ABIS.

An ABI (Application binary interface https://en.wikipedia.org/wiki/Application_binary_interface) is a "language" which a processor/CPU understands.

Most processors/CPUs understand multiple languages (for example you can run 32 bit and 64 bit application on Windows because the processor/CPU understands x86 and amd64. On the newer Macs you need the translation tool Rosetta because the M1 chips don't understand x86)

AFAIK Android devices can use these ABIs/languages (https://github.com/Tobi823/ffupdater/blob/a3a0c2a0550970ca044985b0617f43c16aa64cf6/ffupdater/src/main/java/de/marmaro/krt/ffupdater/device/ABI.kt): 

- arm64-v8a
- armeabi-v7a
- armeabi (ARMv5/6 are old architectures that haven’t been used in years)
- x86
- x86_64
- mips (MIPS is exclusively used for some rare embedded uses of Android)
- mips64 (MIPS is exclusively used for some rare embedded uses of Android)

arm64-v8a, x86_64 and mips64 are 64 bit. 
armeabi-v7a, armeabi, x86 and mips are 32 bit.

Here are some examples which ABIs are supported:

- Android emulator: x86_64, arm64-v8a
- Google Pixel 3a: arm64-v8a, armeabi-v7a, armeabi
- Fairphone 4: arm64-v8a, armeabi-v7a, armeabi
- Amazon Fire Tab Version???: armeabi-v7a, armeabi
- Nexus 7 2013: armeabi-v7a, armeabi
- Samsung SM-P610: arm64-v8a, armeabi-v7a, armeabi

**Comment by Tobi823** [comment]
TLDR: There is no 32 bit or 64 bit Android operating system. If you can run an app depends if the APK file supports at least one of the supported ABIs.

**Comment by Tobi823** [comment]
I hope that helps :)
